teps/0147-tekton-artifacts-phase1.md

```diff
@@ -0,0 +1,410 @@
+- `pkg:github/package-url/purl-spec@244fd47e07d1004`
+- `pkg:docker/cassandra@latest`
+- `pkg:huggingface/distilbert-base-uncased@043235d6088ecd3dd5fb5ca3592b6913fd516027` (ML models)
+- `pkg:maven/org.apache.xmlgraphics/batik-anim@1.9.1`
+
+`purl-spec` also provides a common standard that `Tasks/Steps` can leverage so that output metadata information of one `Task/Step` could be consumed by another in the future. Additionally, it becomes a standard way to distinguish between types of artifacts. [package-url](https://github.com/package-url) also provides client libraries in multiple languages to parse the `purl`. This is something users could leverage.
+
+##### `digest`
+
+The `digest` component includes the result of applying an hash function to the artifact itself.
+The `digest` is in the format `map[string]string` e.g. `{"hash-algorithm":"hex"}`. Example values:
+
+```yaml
+values:
+  uri: "pkg:github/package-url/purl-spec@244fd47e07d1004"
+  digest:
+    sha1: sadhy2d83hd3
+    sha256: ndskajdkasjdlkalsdj39348949
+```
+
+### From Step into Status
+
+The artifact provenance data (i.e. metadata content) is extracted from the local storage using the same mechanism used for results. Results and artifacts data is collated together and stored either in the termination message or in a sidecar log, depending on configuration.
+
+
+The artifact data is then read by the Tekton `TaskRun` controller and copied stored into the `Step State`:
+
+```yaml!
+status:
+  results:
+  - name: release
+    type: string
+    value: |
+      https://storage.googleapis.com/tekton-releases/pipeline/previous/v0.50.2/release.yaml
+  steps:
+    - container: step-1
+      inputs:
+        - name: source
+          value:
+            - uri: pkg:generic/source
+              digest: sha256:8796357729cfd877cf8fa7d45a8ab3524d9249c23a0bf68bb0026c0783b881d2
+  steps:
+    - container: step-2
+      outputs:
+        - name: release-file
+          value:
+            - uri: pkg:generic/release-file
+              digest: sha256:33a06c928729e52d1991a2c55765a7c30ef72b098533f220f3f1d6f352fd32e8
+```
+### From Status into Provenance
+
+The Tekton Chains formatter would access the artifact information from the `StepState`. 
+- All `inputs` will be inserted in the `resolvedDependency` section of the `SLSA provenance`.
+- All `outputs` will be placed as `subjects`.
+  - That's because with just information about the `uri` and `digest`, Tekton Chains does not have enough information to distinguish `subjects` from `byProducts`. Infact, even the `StepAction`/`Task`(for inlined-steps) author does not have the context of how an artifact is used (`byProduct` or `subject` of the build). Without some user input, it is not possible for Chains to determine where to store this information so we propose to fit all `output` artifacts as `subjects` in the `SLSA` provenance.
```

Comment:
If we need to, we can add a byproduct field that acts like output (though it's not immediately obvious to me if we need this right now so I'd hold off on it)

Reply:
I agree. Let's hold off on it until we need it.